### PR TITLE
feat: As a user, I want to have an example schema checker module

### LINKED ISSUE #267: [US] As a user, I want to have an example schema checker module
As a user,
I want to have an example schema checker module,
So that I can easily know how to check the schema that an attestation is based on

----

### Description

The deliverable for this user story is a contract in the `/src/example` folder of the repository that can be used as a module to verify the schema that an incoming attestation is based on.

The module should implement the `AbstractModule` abstract contract and can be based on the `/src/example/CorrectModule.sol` contract.

Most attestation issuers will want to lock their portal down so that it can only accept one, or a specific set, of schemas.  This means they will need to have a module that verifies the schema id in the `AttestationPayload` and checks that it equals a specific schema id.

See here for more info: https://github.com/Consensys/linea-attestation-registry/blob/1be21808f4e41bfa1c1f0c943ee296f4664a7510/contracts/src/types/Structs.sol#L4

As well as that, the module should show an example of decoding the attestation data and checking individual values.

for example of code that does this, see the documentation:

https://verax.gitbook.io/verax/developer-guides/for-attestation-issuers/encoding-attestation-data

### PR DESCRIPTION
**Pull request by Builddddder**
### Related ticket

Fixes #267 

### Type of change

- [ ] Chore
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update

## Check list

- [x] Unit tests for any smart contract change
- [x] Contracts and functions are documented

### CODE REVIEW

**Review by orbmis — APPROVED** [review]

**Comment by orbmis** [comment]
Would it be better to have a hardcoded schema id here? This module is really an example to make it clear to other developers how you would check a schema, but maybe this could be clearer. What do you think?

What about this:

```Solidity
function run(
    AttestationPayload memory attestationPayload,
    bytes memory /*validationPayload*/,
    address /*txSender*/,
    uint256 /*value*/
  ) public view override {
    if (keccak256(attestationPayload.schemaId) != keccak256(abi.encode("HARDCODED_SCHEMA_ID"))) {
      revert InvalidSchemaId();
    }

    abi.decode(attestationPayload.attestationData, (address, bytes32, uint64));
  }
  ```
  
If the payload adheres to the Schema, this will pass, if not, it will revert. Does that make sense?

**Comment by Builddddder** [comment]
I think that's ok for me

**Comment by Builddddder** [comment]
@orbmis Updated code, please review again.

File: contracts/src/examples/modules/SchemaCheckerModule.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { AbstractModule } from "../../interface/AbstractModule.sol";
+import { AttestationPayload } from "../../types/Structs.sol";
+
+/**
+ * @title Schema checker Module
+ * @notice This contract illustrates a valid Module that is used to check the schema that an attestation is based on
+ */
+contract SchemaCheckerModule is AbstractModule {
+  error InvalidSchemaId();
+  struct Profile {
```
Comment (0xEillo) [inline contracts/src/examples/modules/SchemaCheckerModule.sol:13]: I don't understand why there is this struct and we are decoding the attestationData

Comment (Builddddder) [reply contracts/src/examples/modules/SchemaCheckerModule.sol:13]: Just trying to demonstrate how to encode & decode complex struct in the attestation data

Comment (alainncls) [reply contracts/src/examples/modules/SchemaCheckerModule.sol:13]: Maybe we just need to add a comment saying just that: this is an example of how to check the attestation payload encoding/decoding, based on a "Profile Schema"?

Comment (Builddddder) [reply contracts/src/examples/modules/SchemaCheckerModule.sol:16]: Comments had been added @orbmis

**Review by alainncls — APPROVED** [review]

**Pull request merged**